### PR TITLE
Improving MCXRecursive

### PR DESCRIPTION
**Pull request by adjs**
### Summary
Issue #5872 
Change the recursive method for the Lemma 9 of arXiv:1501.06911, first shown in Lemma 7.3 of https://link.aps.org/doi/10.1103/PhysRevA.52.3457

### Details and comments
For the case of num_qubits>5 we changed the recursion method for a method which divides the control qubits in two groups, and them apply the MCXVChain gate alternating between the two groups.

### CODE REVIEW

File: test/python/circuit/test_circuit_qasm.py [hunk 1]
```diff
@@ -270,6 +270,8 @@ def test_circuit_qasm_with_mcx_gate(self):
         qc = QuantumCircuit(4)
         qc.mcx([0, 1, 2], 3)
 
+        print(qc.decompose().draw())
+
```
Comment (adjs) [inline test/python/circuit/test_circuit_qasm.py:274]:
```suggestion
```

Comment (rafaella-vale) [reply test/python/circuit/test_circuit_qasm.py]: Sorry, I forgot.

**Comment by adjs** [comment]
This PR solves issue #9743.

The number of cx gates will be further reduced with PR #9687.

**Review by ShellyGarion — CHANGES_REQUESTED** [review]
Thanks @adjs for your contribution. 
This PR is waiting for #9687 to get merged. 
There are some conflicts that should be fixed afterwards.

Is there some bound on the number of `CX` gates? If so - could you please add it to the docs and add proper tests?

Generally, the name `MCXRecursive` is a bit misleading now, since it's not a recursive algorithm anymore. It basically handles the case of one clean ancilla qubit. We don't want to update the class name now (maybe we will have a unified API of MCX gates towards Qiskit 2.0), but the docstring of the `MCXRecursive` class should be updated to better explain this, also add a reference to the relevant paper and theorem. Perhaps even some examples if possible?
Release notes are also needed.

In addition, in #9687 you introduced two parameters: `relative_phase, action_only` for the `MCXVChain` class. 
Does any of these parameters can also improve the number of `CX` gates in the  `MCXRecursive` class?

**Comment by ShellyGarion** [comment]
Thank you very much for your contribution to Qiskit @adjs! 
Now that #9687 has been merged, could you please rebase your PR on Qiskit main branch, add some release notes and tests, and relate to my comment above?
We are very much interested in merging this PR for Qiskit 1.2 release

**Comment by ShellyGarion** [comment]
I'm closing this PR in favor of #12772, so we'll be able to make it for Qiskit 1.2 release. All your commits appear there as well.
Thank you very much for your contribution to Qiskit!

**Pull request closed without merging**